File: crates/klyron_mysql/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::MysqlBuilder::Impl {
public:
  Klyron::MysqlConfig config;
  bool verbose = false;
};

Klyron::MysqlBuilder::Klyron::MysqlBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::MysqlBuilder& Klyron::MysqlBuilder::withConfig(const Klyron::MysqlConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::MysqlBuilder& Klyron::MysqlBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::MysqlBuilder::Klyron::MysqlInstance Klyron::MysqlBuilder::build() {
  Klyron::MysqlInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_mysql/bindings/rust/builder.rs
//! Builder pattern for klyron_mysql

use crate::types::Klyron::MysqlConfig;

#[derive(Debug, Default)]
pub struct Klyron::MysqlBuilder {
    config: Option<Klyron::MysqlConfig>,
    verbose: bool,
}

impl Klyron::MysqlBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::MysqlConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::MysqlInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::MysqlInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::MysqlInstance {
    pub config: Klyron::MysqlConfig,
    pub verbose: bool,
}
